File: app/src/test/java/com/peterlaurence/trekadvisor/util/ZipTest.java
package com.peterlaurence.trekadvisor.util;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

import java.io.File;
import java.io.IOException;
import java.net.URL;

import static org.junit.Assert.fail;

/**
 * This tests the {@link ZipTask} against the {@link UnzipTask}. A sample map (a simple folder
 * structure) located in the resources of the app is zipped in a temporary folder. Right after
 * that, it's unzipped in the same location. <br>
 * The test is considered successful if this operation is done completely without any error.
 *
 * @author peterLaurence on 10/08/17.
 */
@RunWith(RobolectricTestRunner.class)
public class ZipTest {
    private static final String MAP_NAME = "libvips-with-json";
    private static File mMapsDirectory;

    static {
        try {
            URL mapDirURL = ZipTest.class.getClassLoader().getResource("maps");
            mMapsDirectory = new File(mapDirURL.toURI());
        } catch (Exception e) {
            System.out.println("No resource file for map test directory.");
        }
    }

    private File mTestFolder = new File(System.getProperty("java.io.tmpdir"), "junit_ziptest");

    @Test
    public void zipTest() {
        if (mMapsDirectory != null) {
            final File libVipsMapDir = new File(mMapsDirectory, MAP_NAME);

            final UnzipTask.UnzipProgressionListener unzipProgressionListener = new UnzipTask.UnzipProgressionListener() {
                @Override
                public void onProgress(int p) {

                }

                @Override
                public void onUnzipFinished(File outputDirectory) {
                    System.out.println("Unzip finished");
                    FileUtils.deleteRecursive(mTestFolder);
                }

                @Override
                public void onUnzipError() {
                    fail();
                    FileUtils.deleteRecursive(mTestFolder);
                }
            };

            try {
                final File tempMapArchive = new File(mTestFolder, "testmap.zip");
                tempMapArchive.getParentFile().mkdirs();
                tempMapArchive.createNewFile();

                ZipTask.ZipProgressionListener progressionListener = new ZipTask.ZipProgressionListener() {
                    @Override
                    public void fileListAcquired() {
                        System.out.println("File list acquired");
                    }

                    @Override
                    public void onProgress(int p) {

                    }

                    @Override
                    public void onZipFinished(File outputDirectory) {
                        UnzipTask unzipTask = new UnzipTask(tempMapArchive, mTestFolder, unzipProgressionListener);
                        unzipTask.start();
                    }

                    @Override
                    public void onZipError() {
                        fail();
                    }
                };

                ZipTask zipTask = new ZipTask(libVipsMapDir, tempMapArchive, progressionListener);
                zipTask.execute();

            } catch (IOException e) {
                e.printStackTrace();
            }
        } else {
            fail();
        }
    }
}

File: app/src/test/java/com/peterlaurence/trekme/util/ZipTest.kt
package com.peterlaurence.trekme.util

import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner

import java.io.File
import java.io.IOException

import org.junit.Assert.fail
import java.io.FileInputStream
import java.io.FileOutputStream

/**
 * This tests the [ZipTask] against the [unzipTask].
 *
 * A sample map (a simple folder structure) located in the resources of the app is zipped in a
 * temporary folder. Right after that, it's unzipped in the same location. <br></br>
 * The test is considered successful if this operation is done completely without any error.
 *
 * @author peterLaurence on 10/08/17 -- Converted to Kotlin on 19/05/2019
 */
@RunWith(RobolectricTestRunner::class)
class ZipTest {

    private val mTestFolder = File(System.getProperty("java.io.tmpdir"), "junit_ziptest")

    @Test
    fun zipTest() {
        if (mMapsDirectory != null) {
            val libVipsMapDir = File(mMapsDirectory, MAP_NAME)

            val unzipProgressionListener = object : UnzipProgressionListener {
                override fun onProgress(p: Int) {

                }

                override fun onUnzipFinished(outputDirectory: File) {
                    println("Unzip finished")
                    FileUtils.deleteRecursive(mTestFolder)
                }

                override fun onUnzipError() {
                    fail()
                    FileUtils.deleteRecursive(mTestFolder)
                }
            }

            try {
                val tempMapArchive = File(mTestFolder, "testmap.zip")
                val size = tempMapArchive.length()
                tempMapArchive.parentFile.mkdirs()
                tempMapArchive.createNewFile()

                val progressionListener = object : ZipTask.ZipProgressionListener {
                    override fun fileListAcquired() {
                        println("File list acquired")
                    }

                    override fun onProgress(p: Int) {

                    }

                    override fun onZipFinished() {
                        unzipTask(FileInputStream(tempMapArchive), mTestFolder, size, unzipProgressionListener)
                    }

                    override fun onZipError() {
                        fail()
                    }
                }

                val zipTask = ZipTask(libVipsMapDir, FileOutputStream(tempMapArchive), progressionListener)
                zipTask.execute()

            } catch (e: IOException) {
                e.printStackTrace()
            }

        } else {
            fail()
        }
    }

    companion object {
        private const val MAP_NAME = "libvips-with-json"
        private var mMapsDirectory: File? = null

        init {
            try {
                val mapDirURL = ZipTest::class.java.classLoader!!.getResource("maps")
                mMapsDirectory = File(mapDirURL.toURI())
            } catch (e: Exception) {
                println("No resource file for map test directory.")
            }
        }
    }
}
